BUG: Ensure `__array_finalize__` cannot back-mangle shape

mhvk opened the pull request:
Just a small step, but `__array_finalize__` should not be allowed to change the shape of any `obj` passed in. Arguably, it should not change the shape of the mask at all, but removing that currently breaks quite a bit more, so perhaps it is best to try to remedy that in a second step.

Need to think a bit about a test case that shows that this actually prevents problems.

Commit:
BUG: Ensure __array_finalize__ cannot back-mangle shape

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -3351,6 +3354,8 @@ def __setattr__(self, attr, value):
                 self._mask.shape = self.shape
             except (AttributeError, TypeError):
                 pass
+        elif attr == 'shape' and getmask(self) is not nomask:
```
Might be clearer to add `dtype` and `shape` `@properties` that fall back on `ndarray.shape.__get__` etc

eric-wieser commented on numpy/ma/tests/test_old_ma.py:
```diff
@@ -273,7 +273,8 @@ def test_testCopySize(self):
         y1a = array(y1, copy=0)
-        assert_(y1a.mask is y1.mask)
+        assert_(y1a._mask.__array_interface__ ==
+                y1._mask.__array_interface__)
```
Hmmm... I'd actually expect the stronger result of `y1a is y1` here

eric-wieser replied:
This is #4043, I think

eric-wieser replied:
Maybe add a comment referencing that issue here?

eric-wieser commented on numpy/ma/tests/test_core.py:
```diff
@@ -397,7 +397,8 @@ def test_copy(self):
         y1a = array(y1)
         assert_(y1a._data.__array_interface__ ==
```
Wait, what? This doesn't seem sensible to me at all (#10318)

mhvk commented:
@eric-wieser - I agree the changes in behaviour for `np.ma.array` are not good, and I probably should work around them in the initializer.

Also agreed that setters for `shape` and `dtype` would be better than overriding setting for all attributes with `__setattr__`. I wasn't quite sure how to properly do this for properties - but a bit of googling suggests one has to do something like `super().shape.__set__(<shape>)`.

mhvk committed:
MAINT: Replace ma.__setattr__ with dtype and shape overrides.

Since dtype and shape are properties, this needs a somewhat
ugly super construction; see https://bugs.python.org/issue14965

mhvk commented:
@eric-wieser - your comment reminded me about the existence of this branch...  I now added a second commit replacing `__setattr__` with `shape` and `dtype` properties and setters. I'm not sure I like it that much more... If you are also in doubt, perhaps I should just remove it so this one can focus on just the fix for the back-mangling of shapes.

I also still need to think of a test case that shows this actually prevents any problems!

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -3344,17 +3346,33 @@ def __setitem__(self, indx, value):
+
+    @shape.setter
+    def shape(self, shape):
+        super(MaskedArray, MaskedArray).shape.__set__(self, shape)
```
I think this should be just `super(MaskedArray).shape` or perhaps just `ndarray.shape`

eric-wieser replied:
Edit: I'm wrong. TIL how `super(cls, cls)` works

mhvk replied:
Yes, this annoyance is in part why I wondered if it was a good idea to replace `__setattr__` with this...

eric-wieser replied:
The advantage is that subclasses are now able to call `Maskedarray.shape.__set__` in the unlikely even they override shape, and it won't accidentally call `ndarray.shape.__set__`

eric-wieser replied:
Actually, I think this should be `super(MaskedArray, type(self))`

eric-wieser commented:
Is this enough for a test? On `master`:

```
>>> d = a.data
>>> m = a.mask
>>> a.shape = (2, 2)
>>> d
array([1, 2, 3, 4])
>>> m  # your patch stops this changing shape, which is consistent with d
array([[ True,  True],
       [ True,  True]], dtype=bool)
```

mhvk commented:
Hmm, that particular case is not solved by this PR - it needs your original one, where the mask is always viewed.  Indeed, I found it very hard to far to get to a case where my PR matters...  I'll try a bit more, but if it fails, perhaps best to split it into two - and at least get rid of the `__setattr__` override.

eric-wieser commented:
Ok, we now have a concrete problem caused by this, as documented in #3140

mhvk commented:
Thanks for realizing there was a test case after all! I now added it, and rebased for good measure (and fixed the `super` following your comment).

eric-wieser commented on numpy/ma/core.py:
```diff
@@ -3344,17 +3346,35 @@ def __setitem__(self, indx, value):
 
+    @property
+    def shape(self):
+        return super(MaskedArray, self).shape
```
Pretty sure this needs to be `super(MaskedArray, type(self)).__get__(self, type(self))`, unless there's something weird about base classes in C

eric-wieser replied:
Ok, seems there's something weird about base classes. If this were shadowing another `property` (rather than a `getset_descriptor`) , you'd need what I have above.

mhvk replied:
I don't think that's true - I;m fairly sure one can shadow a property getter as I have it above, just not the setter.

eric-wieser commented:
Thanks @mhvk!
